File: demos/UnityModuleTest/Packages/Xrpa/Runtime/utils/SpmcRingBuffer.cs
/*
// @generated
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;

namespace Xrpa
{
    /// <summary>
    /// Lock-free Single Producer Multiple Consumer (SPMC) ring buffer.
    ///
    /// Memory Layout:
    /// Offset  | Size | Field
    /// --------|------|------------------
    /// 0       | 4    | poolSize (total memory after header)
    /// 4       | 4    | blockSize (aligned size of each block, includes 4-byte header)
    /// 8       | 4    | blockCount (number of blocks)
    /// 12      | 4    | (padding/reserved)
    /// 16      | 4    | writeIndex (atomic uint32, monotonically increasing)
    /// 20      | 4    | minReadIndex (atomic uint32, minimum valid read index)
    /// 24      | ...  | Block pool starts here
    ///
    /// Block Layout:
    /// Offset  | Size | Field
    /// --------|------|------------------
    /// 0       | 4    | dataSize (0 = skipped block, greater than 0 = total data size in bytes)
    /// 4       | ...  | data (up to blockSize - 4 bytes per block)
    ///
    /// Multi-block entries:
    /// - First block header contains total data size across all blocks
    /// - Subsequent blocks have NO header (data is contiguous across blocks)
    /// - If entry would wrap, remaining blocks at end are marked as skipped (size=0)
    ///   and entry starts at block 0
    /// </summary>
    unsafe public class SpmcRingBuffer
    {
        public static readonly int BLOCK_HEADER_SIZE = 4;
        public static readonly int HEADER_SIZE = 24;

        private MemoryAccessor _memSource;
        private MemoryAccessor _headerAccessor;
        private MemoryAccessor _poolAccessor;

        private int _blockSize = 0;
        private int _blockCount = 0;

        public SpmcRingBuffer() { }

        /// <summary>
        /// Constructor takes MemoryAccessor + offset (matches C++ and Python pattern)
        /// </summary>
        public SpmcRingBuffer(MemoryAccessor memSource, long memOffset)
        {
            _memSource = memSource.Slice(memOffset);
            _headerAccessor = memSource.Slice(memOffset, HEADER_SIZE);

            if (!_headerAccessor.IsNull())
            {
                MemoryOffset pos = new();
                int poolSize = _headerAccessor.ReadInt(pos);
                _blockSize = _headerAccessor.ReadInt(pos);
                _blockCount = _headerAccessor.ReadInt(pos);

                if (poolSize > 0 && _blockSize > 0 && _blockCount > 0)
                {
                    _poolAccessor = memSource.Slice(memOffset + HEADER_SIZE, poolSize);
                }
            }
        }

        /// <summary>
        /// Calculate total memory size needed for a ring buffer with given block size and count.
        /// </summary>
        public static int GetMemSize(int blockSize, int blockCount)
        {
            blockSize = Align(blockSize);
            return HEADER_SIZE + (blockSize * blockCount);
        }

        /// <summary>
        /// Initialize the ring buffer with given block size and count.
        /// </summary>
        public void Init(int blockSize, int blockCount)
        {
            blockSize = Align(blockSize);
            _blockSize = blockSize;
            _blockCount = blockCount;

            int poolSize = blockSize * blockCount;

            MemoryOffset pos = new();
            _headerAccessor.WriteInt(poolSize, pos);    // poolSize
            _headerAccessor.WriteInt(blockSize, pos);   // blockSize
            _headerAccessor.WriteInt(blockCount, pos);  // blockCount
            _headerAccessor.WriteInt(0, pos);           // reserved
            _headerAccessor.WriteUint(0, pos);          // writeIndex
            _headerAccessor.WriteUint(0, pos);          // minReadIndex

            _poolAccessor = _memSource.Slice(HEADER_SIZE, poolSize);
        }

        public bool IsNull()
        {
            return _headerAccessor == null || _headerAccessor.IsNull() ||
                   _poolAccessor == null || _poolAccessor.IsNull();
        }

        public int BlockSize => _blockSize;
        public int BlockCount => _blockCount;

        /// <summary>
        /// Maximum data that can fit in a single entry (all blocks - one header)
        /// </summary>
        public int MaxDataSize => (_blockSize * _blockCount) - BLOCK_HEADER_SIZE;

        /// <summary>
        /// Producer: write data to the ring buffer.
        /// Returns true if write succeeded.
        /// Calls callback with MemoryAccessor for writing the data.
        /// </summary>
        public bool Write(int dataSize, Action<MemoryAccessor> callback)
        {
            if (IsNull() || dataSize <= 0)
            {
                return false;
            }

            int blocksNeeded = GetBlocksNeeded(dataSize);
            if (blocksNeeded > _blockCount)
            {
                return false;
            }

            uint writeIndex = LoadWriteIndex();
            uint startBlockIndex = writeIndex % (uint)_blockCount;

            // Check if entry would wrap
            uint endBlockIndex = startBlockIndex + (uint)blocksNeeded;
            uint newWriteIndex;
            int skippedBlocks = 0;

            if (endBlockIndex > (uint)_blockCount)
            {
                // Need to wrap - mark remaining blocks as skipped
                skippedBlocks = _blockCount - (int)startBlockIndex;
                startBlockIndex = 0;
                newWriteIndex = writeIndex + (uint)(skippedBlocks + blocksNeeded);
            }
            else
            {
                newWriteIndex = writeIndex + (uint)blocksNeeded;
            }

            // Update minReadIndex to make room (evict old blocks)
            uint minReadIndex = LoadMinReadIndex();
            uint requiredMinReadIndex = newWriteIndex > (uint)_blockCount
                ? newWriteIndex - (uint)_blockCount
                : 0;

            if (minReadIndex < requiredMinReadIndex)
            {
                // Need to advance minReadIndex to a valid start block
                // Walk through entries starting from current minReadIndex until we pass required
                uint newMinReadIndex = SkipToValidEntry(minReadIndex, requiredMinReadIndex);
                StoreMinReadIndex(newMinReadIndex);
            }

            // Mark skipped blocks (if wrapping)
            if (skippedBlocks > 0)
            {
                for (int i = 0; i < skippedBlocks; ++i)
                {
                    int blockOffset = GetBlockOffset((writeIndex + (uint)i) % (uint)_blockCount);
                    SetBlockDataSize(blockOffset, 0);
                }
            }

            // Write data size to the header of the first block
            int firstBlockOffset = GetBlockOffset(startBlockIndex);
            SetBlockDataSize(firstBlockOffset, (uint)dataSize);

            // Create MemoryAccessor for the data region
            // Data starts after the first block's header and spans blocksNeeded blocks
            int dataOffset = firstBlockOffset + BLOCK_HEADER_SIZE;
            int maxDataSpace = (blocksNeeded * _blockSize) - BLOCK_HEADER_SIZE;
            MemoryAccessor dataAccessor = _poolAccessor.Slice(dataOffset, maxDataSpace);

            callback(dataAccessor);

            // Update writeIndex after data is written
            StoreWriteIndex(newWriteIndex);

            return true;
        }

        private static int Align(int x)
        {
            return (x + 3) & ~3;
        }

        private int GetBlockOffset(uint blockIndex)
        {
            return (int)blockIndex * _blockSize;
        }

        internal int GetBlocksNeeded(int dataSize)
        {
            // First block holds dataSize bytes - BLOCK_HEADER_SIZE
            // Subsequent blocks hold blockSize bytes each
            int firstBlockData = _blockSize - BLOCK_HEADER_SIZE;
            if (dataSize <= firstBlockData)
            {
                return 1;
            }
            int remaining = dataSize - firstBlockData;
            // Use (remaining - 1) / _blockSize + 1 to avoid overflow from remaining + _blockSize - 1
            return 1 + ((remaining - 1) / _blockSize) + 1;
        }

        private uint* GetWriteIndexPtr()
        {
            return (uint*)(_headerAccessor.DataSource + _headerAccessor.MemOffset + 16);
        }

        private uint* GetMinReadIndexPtr()
        {
            return (uint*)(_headerAccessor.DataSource + _headerAccessor.MemOffset + 20);
        }

        internal uint LoadWriteIndex()
        {
            return AtomicUtils.AtomicLoadAcquire(GetWriteIndexPtr());
        }

        internal uint LoadMinReadIndex()
        {
            return AtomicUtils.AtomicLoadAcquire(GetMinReadIndexPtr());
        }

        private void StoreWriteIndex(uint value)
        {
            AtomicUtils.AtomicStoreRelease(GetWriteIndexPtr(), value);
        }

        private void StoreMinReadIndex(uint value)
        {
            AtomicUtils.AtomicStoreRelease(GetMinReadIndexPtr(), value);
        }

        private void SetBlockDataSize(int blockOffset, uint dataSize)
        {
            XrpaUtils.BoundsAssert(blockOffset, BLOCK_HEADER_SIZE, 0, _poolAccessor.Size);
            *((uint*)(_poolAccessor.DataSource + _poolAccessor.MemOffset + blockOffset)) = dataSize;
        }

        internal uint GetBlockDataSize(int blockOffset)
        {
            XrpaUtils.BoundsAssert(blockOffset, BLOCK_HEADER_SIZE, 0, _poolAccessor.Size);
            return *((uint*)(_poolAccessor.DataSource + _poolAccessor.MemOffset + blockOffset));
        }

        /// <summary>
        /// Skip to a valid start block (one with dataSize greater than 0)
        /// </summary>
        private uint SkipToValidBlock(uint startIndex)
        {
            uint writeIndex = LoadWriteIndex();
            while (startIndex < writeIndex)
            {
                uint blockIndex = startIndex % (uint)_blockCount;
                int blockOffset = GetBlockOffset(blockIndex);
                uint dataSize = GetBlockDataSize(blockOffset);
                if (dataSize > 0)
                {
                    return startIndex;
                }
                startIndex++;
            }
            return startIndex;
        }

        /// <summary>
        /// Walk through entries starting from currentIndex until we find one at or past targetIndex.
        /// This properly handles multi-block entries by advancing by blocksNeeded instead of 1.
        /// </summary>
        private uint SkipToValidEntry(uint currentIndex, uint targetIndex)
        {
            uint writeIndex = LoadWriteIndex();

            while (currentIndex < writeIndex && currentIndex < targetIndex)
            {
                uint blockIndex = currentIndex % (uint)_blockCount;
                int blockOffset = GetBlockOffset(blockIndex);
                uint dataSize = GetBlockDataSize(blockOffset);

                if (dataSize == 0)
                {
                    // Skipped block (from wrapping), advance by 1
                    currentIndex++;
                }
                else
                {
                    // Valid entry - skip past all its blocks
                    int blocksNeeded = GetBlocksNeeded((int)dataSize);
                    currentIndex += (uint)blocksNeeded;
                }
            }

            // Now currentIndex >= targetIndex (or at writeIndex if no more entries)
            // Make sure we're at a valid entry start, not in the middle of skipped blocks
            return SkipToValidBlock(currentIndex);
        }

        internal MemoryAccessor PoolAccessor => _poolAccessor;
    }

    /// <summary>
    /// Iterator for reading from an SPMC ring buffer.
    /// Each consumer should maintain its own iterator.
    /// </summary>
    public class SpmcRingBufferIterator
    {
        private uint _localReadIndex = 0;

        /// <summary>
        /// Check if entries were missed (overwritten before reading)
        /// </summary>
        public bool HasMissedEntries(SpmcRingBuffer ringBuffer)
        {
            uint minReadIndex = ringBuffer.LoadMinReadIndex();
            return _localReadIndex < minReadIndex;
        }

        /// <summary>
        /// Check if there are unread entries
        /// </summary>
        public bool HasNext(SpmcRingBuffer ringBuffer)
        {
            uint writeIndex = ringBuffer.LoadWriteIndex();
            return _localReadIndex < writeIndex;
        }

        /// <summary>
        /// Read next entry; callback receives MemoryAccessor.
        /// Returns false if data was overwritten during read (stale read).
        /// </summary>
        public bool ReadNext(SpmcRingBuffer ringBuffer, Action<MemoryAccessor> callback)
        {
            if (!HasNext(ringBuffer))
            {
                return false;
            }

            // Skip to valid block if we fell behind
            if (HasMissedEntries(ringBuffer))
            {
                _localReadIndex = ringBuffer.LoadMinReadIndex();
            }

            // Skip any skipped blocks (size=0)
            uint writeIndex = ringBuffer.LoadWriteIndex();
            while (_localReadIndex < writeIndex)
            {
                uint blockIndex = _localReadIndex % (uint)ringBuffer.BlockCount;
                int blockOffset = (int)blockIndex * ringBuffer.BlockSize;
                uint dataSize = ringBuffer.GetBlockDataSize(blockOffset);

                if (dataSize == 0)
                {
                    // Skipped block, advance
                    _localReadIndex++;
                    continue;
                }

                // Found a valid entry
                int blocksNeeded = ringBuffer.GetBlocksNeeded((int)dataSize);

                // Create MemoryAccessor for the data
                int dataOffset = blockOffset + SpmcRingBuffer.BLOCK_HEADER_SIZE;
                int maxDataSpace = (blocksNeeded * ringBuffer.BlockSize) - SpmcRingBuffer.BLOCK_HEADER_SIZE;
                MemoryAccessor dataAccessor = ringBuffer.PoolAccessor.Slice(dataOffset, maxDataSpace);

                callback(dataAccessor);

                // Validate that data is still valid after read
                uint newMinReadIndex = ringBuffer.LoadMinReadIndex();
                if (_localReadIndex < newMinReadIndex)
                {
                    // Data was overwritten during read - stale read
                    _localReadIndex = newMinReadIndex;
                    return false;
                }

                // Advance past this entry
                _localReadIndex += (uint)blocksNeeded;
                return true;
            }

            return false;
        }

        /// <summary>
        /// Skip to current write position
        /// </summary>
        public void SetToEnd(SpmcRingBuffer ringBuffer)
        {
            _localReadIndex = ringBuffer.LoadWriteIndex();
        }
    }
}

File: modules/xred-debug/ImageViewer/xrpa_runtime/utils/spmc_ring_buffer.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# @generated
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Lock-free Single Producer Multiple Consumer (SPMC) ring buffer.

Memory Layout:
Offset  | Size | Field
--------|------|------------------
0       | 4    | poolSize (total memory after header)
4       | 4    | blockSize (aligned size of each block, includes 4-byte header)
8       | 4    | blockCount (number of blocks)
12      | 4    | (padding/reserved)
16      | 4    | writeIndex (atomic uint32, monotonically increasing)
20      | 4    | minReadIndex (atomic uint32, minimum valid read index)
24      | ...  | Block pool starts here

Block Layout:
Offset  | Size | Field
--------|------|------------------
0       | 4    | dataSize (0 = skipped block, >0 = total data size in bytes)
4       | ...  | data (up to blockSize - 4 bytes per block)

Multi-block entries:
- First block header contains total data size across all blocks
- Subsequent blocks have NO header (data is contiguous across blocks)
- If entry would wrap, remaining blocks at end are marked as skipped (size=0)
  and entry starts at block 0
"""

import warnings
from typing import Callable, Optional

from xrpa_runtime.utils.memory_accessor import MemoryAccessor, MemoryOffset
from xrpa_runtime.utils.xrpa_utils import xrpa_debug_bounds_assert

# Try to import C extension for true atomic operations
try:
    import xrpa_atomics

    _HAS_ATOMICS = True
except ImportError:
    _HAS_ATOMICS = False
    warnings.warn(
        "xrpa_atomics C extension not available. "
        "SPMC ring buffer will use non-atomic operations which may not be safe "
        "for cross-process/cross-thread shared memory.",
        RuntimeWarning,
        stacklevel=2,
    )


def _align(x: int) -> int:
    """Align x to 4-byte boundary."""
    return (x + 3) & ~3


def _atomic_load_acquire(data_source: memoryview, offset: int) -> int:
    """Atomically load a uint32 with acquire semantics."""
    if _HAS_ATOMICS:
        return xrpa_atomics.atomic_load_acquire(data_source, offset)
    else:
        # Fallback: non-atomic read (not safe for shared memory)
        import struct

        return struct.unpack("<I", data_source[offset : offset + 4])[0]


def _atomic_store_release(data_source: memoryview, offset: int, value: int) -> None:
    """Atomically store a uint32 with release semantics."""
    if _HAS_ATOMICS:
        xrpa_atomics.atomic_store_release(data_source, offset, value)
    else:
        # Fallback: non-atomic write (not safe for shared memory)
        import struct

        data_source[offset : offset + 4] = struct.pack("<I", value)


class SpmcRingBuffer:
    """Lock-free Single Producer Multiple Consumer (SPMC) ring buffer."""

    BLOCK_HEADER_SIZE = 4
    HEADER_SIZE = 24

    # Header field offsets
    _POOL_SIZE_OFFSET = 0
    _BLOCK_SIZE_OFFSET = 4
    _BLOCK_COUNT_OFFSET = 8
    _RESERVED_OFFSET = 12
    _WRITE_INDEX_OFFSET = 16
    _MIN_READ_INDEX_OFFSET = 20

    def __init__(self, mem_source: MemoryAccessor, mem_offset: int):
        """
        Constructor takes MemoryAccessor + offset (matches C++ and C# pattern).

        Args:
            mem_source: MemoryAccessor providing access to the underlying memory
            mem_offset: Byte offset within mem_source where the ring buffer starts
        """
        self._mem_source = mem_source.slice(mem_offset)
        self._header_accessor = mem_source.slice(mem_offset, self.HEADER_SIZE)
        self._pool_accessor: Optional[MemoryAccessor] = None

        # Cached header values
        self._block_size = 0
        self._block_count = 0

        if not self._header_accessor.is_null():
            pos = MemoryOffset(0)
            pool_size = self._header_accessor.read_int(pos)
            self._block_size = self._header_accessor.read_int(pos)
            self._block_count = self._header_accessor.read_int(pos)

            if pool_size > 0 and self._block_size > 0 and self._block_count > 0:
                self._pool_accessor = mem_source.slice(
                    mem_offset + self.HEADER_SIZE, pool_size
                )

    @staticmethod
    def get_mem_size(block_size: int, block_count: int) -> int:
        """Calculate total memory size needed for a ring buffer."""
        block_size = _align(block_size)
        return SpmcRingBuffer.HEADER_SIZE + (block_size * block_count)

    def init(self, block_size: int, block_count: int) -> None:
        """Initialize the ring buffer with given block size and count."""
        block_size = _align(block_size)
        self._block_size = block_size
        self._block_count = block_count

        pool_size = block_size * block_count

        pos = MemoryOffset(0)
        self._header_accessor.write_int(pool_size, pos)  # poolSize
        self._header_accessor.write_int(block_size, pos)  # blockSize
        self._header_accessor.write_int(block_count, pos)  # blockCount
        self._header_accessor.write_int(0, pos)  # reserved
        self._header_accessor.write_uint(0, pos)  # writeIndex
        self._header_accessor.write_uint(0, pos)  # minReadIndex

        self._pool_accessor = self._mem_source.slice(self.HEADER_SIZE, pool_size)

    def is_null(self) -> bool:
        """Check if the ring buffer is properly initialized."""
        return (
            self._header_accessor is None
            or self._header_accessor.is_null()
            or self._pool_accessor is None
            or self._pool_accessor.is_null()
        )

    @property
    def block_size(self) -> int:
        """Get the aligned block size."""
        return self._block_size

    @property
    def block_count(self) -> int:
        """Get the number of blocks."""
        return self._block_count

    @property
    def max_data_size(self) -> int:
        """Maximum data that can fit in a single entry (all blocks - one header)."""
        return (self._block_size * self._block_count) - self.BLOCK_HEADER_SIZE

    def write(self, data_size: int, callback: Callable[[MemoryAccessor], None]) -> bool:
        """
        Producer: write data to the ring buffer.

        Args:
            data_size: Size of data to write in bytes
            callback: Function called with MemoryAccessor for writing the data

        Returns:
            True if write succeeded, False otherwise
        """
        if self.is_null() or data_size <= 0:
            return False

        blocks_needed = self._get_blocks_needed(data_size)
        if blocks_needed > self._block_count:
            return False

        write_index = self._load_write_index()
        start_block_index = write_index % self._block_count

        # Check if entry would wrap
        end_block_index = start_block_index + blocks_needed
        skipped_blocks = 0

        if end_block_index > self._block_count:
            # Need to wrap - mark remaining blocks as skipped
            skipped_blocks = self._block_count - start_block_index
            start_block_index = 0
            new_write_index = write_index + skipped_blocks + blocks_needed
        else:
            new_write_index = write_index + blocks_needed

        # Update minReadIndex to make room (evict old blocks)
        min_read_index = self._load_min_read_index()
        required_min_read_index = (
            new_write_index - self._block_count
            if new_write_index > self._block_count
            else 0
        )

        if min_read_index < required_min_read_index:
            # Need to advance minReadIndex to a valid start block
            # Walk through entries starting from current minReadIndex until we pass required
            new_min_read_index = self._skip_to_valid_entry(
                min_read_index, required_min_read_index
            )
            self._store_min_read_index(new_min_read_index)

        # Mark skipped blocks (if wrapping)
        if skipped_blocks > 0:
            for i in range(skipped_blocks):
                block_offset = self._get_block_offset(
                    (write_index + i) % self._block_count
                )
                self._set_block_data_size(block_offset, 0)

        # Write data size to the header of the first block
        first_block_offset = self._get_block_offset(start_block_index)
        self._set_block_data_size(first_block_offset, data_size)

        # Create MemoryAccessor for the data region
        # Data starts after the first block's header and spans blocks_needed blocks
        data_offset = first_block_offset + self.BLOCK_HEADER_SIZE
        max_data_space = (blocks_needed * self._block_size) - self.BLOCK_HEADER_SIZE
        data_accessor = self._pool_accessor.slice(data_offset, max_data_space)

        callback(data_accessor)

        # Update writeIndex after data is written
        self._store_write_index(new_write_index)

        return True

    def _get_block_offset(self, block_index: int) -> int:
        """Get byte offset of a block within the pool."""
        return block_index * self._block_size

    def _get_blocks_needed(self, data_size: int) -> int:
        """Calculate number of blocks needed for given data size."""
        # First block holds data_size bytes - BLOCK_HEADER_SIZE
        # Subsequent blocks hold block_size bytes each
        first_block_data = self._block_size - self.BLOCK_HEADER_SIZE
        if data_size <= first_block_data:
            return 1
        remaining = data_size - first_block_data
        # Use (remaining - 1) // block_size + 1 to avoid overflow from remaining + block_size - 1
        return 1 + ((remaining - 1) // self._block_size) + 1

    def _load_write_index(self) -> int:
        """Atomically load the write index with acquire semantics."""
        return _atomic_load_acquire(
            self._header_accessor.data_source,
            self._header_accessor.mem_offset + self._WRITE_INDEX_OFFSET,
        )

    def _load_min_read_index(self) -> int:
        """Atomically load the min read index with acquire semantics."""
        return _atomic_load_acquire(
            self._header_accessor.data_source,
            self._header_accessor.mem_offset + self._MIN_READ_INDEX_OFFSET,
        )

    def _store_write_index(self, value: int) -> None:
        """Atomically store the write index with release semantics."""
        _atomic_store_release(
            self._header_accessor.data_source,
            self._header_accessor.mem_offset + self._WRITE_INDEX_OFFSET,
            value,
        )

    def _store_min_read_index(self, value: int) -> None:
        """Atomically store the min read index with release semantics."""
        _atomic_store_release(
            self._header_accessor.data_source,
            self._header_accessor.mem_offset + self._MIN_READ_INDEX_OFFSET,
            value,
        )

    def _set_block_data_size(self, block_offset: int, data_size: int) -> None:
        """Set the data size in a block header."""
        xrpa_debug_bounds_assert(
            block_offset, self.BLOCK_HEADER_SIZE, self._pool_accessor.size
        )
        self._pool_accessor.write_uint(data_size, MemoryOffset(block_offset))

    def _get_block_data_size(self, block_offset: int) -> int:
        """Get the data size from a block header."""
        xrpa_debug_bounds_assert(
            block_offset, self.BLOCK_HEADER_SIZE, self._pool_accessor.size
        )
        return self._pool_accessor.read_uint(MemoryOffset(block_offset))

    def _skip_to_valid_block(self, start_index: int) -> int:
        """Skip to a valid start block (one with dataSize > 0)."""
        write_index = self._load_write_index()
        while start_index < write_index:
            block_index = start_index % self._block_count
            block_offset = self._get_block_offset(block_index)
            data_size = self._get_block_data_size(block_offset)
            if data_size > 0:
                return start_index
            start_index += 1
        return start_index

    def _skip_to_valid_entry(self, current_index: int, target_index: int) -> int:
        """
        Walk through entries starting from current_index until we find one at or past target_index.
        This properly handles multi-block entries by advancing by blocks_needed instead of 1.
        """
        write_index = self._load_write_index()

        while current_index < write_index and current_index < target_index:
            block_index = current_index % self._block_count
            block_offset = self._get_block_offset(block_index)
            data_size = self._get_block_data_size(block_offset)

            if data_size == 0:
                # Skipped block (from wrapping), advance by 1
                current_index += 1
            else:
                # Valid entry - skip past all its blocks
                blocks_needed = self._get_blocks_needed(data_size)
                current_index += blocks_needed

        # Now current_index >= target_index (or at write_index if no more entries)
        # Make sure we're at a valid entry start, not in the middle of skipped blocks
        return self._skip_to_valid_block(current_index)

    # Internal accessors for iterator
    @property
    def _pool(self) -> MemoryAccessor:
        return self._pool_accessor


class SpmcRingBufferIterator:
    """
    Iterator for reading from an SPMC ring buffer.
    Each consumer should maintain its own iterator.
    """

    def __init__(self):
        self._local_read_index: int = 0

    def has_missed_entries(self, ring_buffer: SpmcRingBuffer) -> bool:
        """Check if entries were missed (overwritten before reading)."""
        min_read_index = ring_buffer._load_min_read_index()
        return self._local_read_index < min_read_index

    def has_next(self, ring_buffer: SpmcRingBuffer) -> bool:
        """Check if there are unread entries."""
        write_index = ring_buffer._load_write_index()
        return self._local_read_index < write_index

    def read_next(
        self, ring_buffer: SpmcRingBuffer, callback: Callable[[MemoryAccessor], None]
    ) -> bool:
        """
        Read next entry; callback receives MemoryAccessor.

        Args:
            ring_buffer: The ring buffer to read from
            callback: Function called with MemoryAccessor containing the data

        Returns:
            False if data was overwritten during read (stale read), True otherwise
        """
        if not self.has_next(ring_buffer):
            return False

        # Skip to valid block if we fell behind
        if self.has_missed_entries(ring_buffer):
            self._local_read_index = ring_buffer._load_min_read_index()

        # Skip any skipped blocks (size=0)
        write_index = ring_buffer._load_write_index()
        while self._local_read_index < write_index:
            block_index = self._local_read_index % ring_buffer.block_count
            block_offset = block_index * ring_buffer.block_size
            data_size = ring_buffer._get_block_data_size(block_offset)

            if data_size == 0:
                # Skipped block, advance
                self._local_read_index += 1
                continue

            # Found a valid entry
            blocks_needed = ring_buffer._get_blocks_needed(data_size)

            # Create MemoryAccessor for the data
            data_offset = block_offset + SpmcRingBuffer.BLOCK_HEADER_SIZE
            max_data_space = (
                blocks_needed * ring_buffer.block_size
            ) - SpmcRingBuffer.BLOCK_HEADER_SIZE
            data_accessor = ring_buffer._pool.slice(data_offset, max_data_space)

            callback(data_accessor)

            # Validate that data is still valid after read
            new_min_read_index = ring_buffer._load_min_read_index()
            if self._local_read_index < new_min_read_index:
                # Data was overwritten during read - stale read
                self._local_read_index = new_min_read_index
                return False

            # Advance past this entry
            self._local_read_index += blocks_needed
            return True

        return False

    def set_to_end(self, ring_buffer: SpmcRingBuffer) -> None:
        """Skip to current write position."""
        self._local_read_index = ring_buffer._load_write_index()
